pkg/provider/doppler/doppler.go

```diff
@@ -0,0 +1,305 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package doppler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/find"
+	"github.com/external-secrets/external-secrets/pkg/provider/doppler/client"
+	"github.com/external-secrets/external-secrets/pkg/utils"
+)
+
+const (
+	customBaseURLEnvVar                                = "DOPPLER_BASE_URL"
+	verifyTLSOverrideEnvVar                            = "DOPPLER_VERIFY_TLS"
+	secretsDownloadFileKey                             = "DOPPLER_SECRETS_FILE"
+	errDopplerStore                                    = "missing or invalid Doppler SecretStore"
+	errDopplerTokenSecretName                          = "missing auth.secretRef.dopplerToken.name"
+	errInvalidClusterStoreMissingDopplerTokenNamespace = "missing auth.secretRef.dopplerToken.namespace"
+	errFetchDopplerTokenSecret                         = "unable to find find DopplerToken secret: %w"
+	errMissingDopplerToken                             = "auth.secretRef.dopplerToken.key '%s' not found in secret '%s'"
+	errNewClient                                       = "unable to create DopplerClient : %s"
+	errInvalidStore                                    = "invalid store: %s"
+	errGetSecret                                       = "could not get secret %s: %s"
+	errGetSecrets                                      = "could not get secrets %s"
+	errUnmarshalSecretMap                              = "unable to unmarshal secret %s: %w"
+)
+
+// https://github.com/external-secrets/external-secrets/issues/644
+var _ esv1beta1.SecretsClient = &Doppler{}
+var _ esv1beta1.Provider = &Doppler{}
+
+type Doppler struct {
+	client          SecretsClientInterface
+	project         string
+	config          string
+	nameTransformer string
+	format          string
+}
+
+// SecretsClientInterface is the required subset of the Doppler API Client.
+type SecretsClientInterface interface {
```

Comment:
Used so this interface can be mocked for unit testing. Standard design pattern used by most providers (GCP).